24.0.0.1-beta staging (#3481)
OpenLiberty/blogs: posts/2023-12-19-24.0.0.1-beta.adoc

---
layout: post

> title: "Jakarta Data 1.0 hits Milestone 2 in 24.0.0.1 beta"

Reviewer: ```suggestion
title: "Jakarta Data 1.0 hits Milestone 2 in 24.0.0.1-beta"
```

# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/dmuelle
author_github: https://github.com/dmuelle

> seo-title: Jakarta Data 1.0 hits Milestone 2 in 24.0.0.1 beta - OpenLiberty.io

Reviewer: ```suggestion
seo-title: Jakarta Data 1.0 hits Milestone 2 in 24.0.0.1-beta - OpenLiberty.io
```

seo-description: Open Liberty 24.0.0.1-beta includes a preview of the new Jakarta Data specification for Jakarta EE, as it currently stands at Milestone 2. Milestone 2 provides API updates to pagination and various improvements to the Javadoc and specification text. You can try it out and give feedback on the specification so far.
blog_description: Open Liberty 24.0.0.1-beta includes a preview of the new Jakarta Data specification for Jakarta EE, as it currently stands at Milestone 2. Milestone 2 provides API updates to pagination and various improvements to the Javadoc and specification text. You can try it out and give feedback on the specification so far.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---

> = Jakarta Data 1.0 hits Milestone 2 in 24.0.0.1 beta

Reviewer: ```suggestion
= Jakarta Data 1.0 hits Milestone 2 in 24.0.0.1-beta
```

See also link:{url-prefix}/blog/?search=beta&key=tag[previous Open Liberty beta blog posts].

[#milestone_2]
== Preview of Jakarta Data (Milestone 2)

Jakarta Data is a new Jakarta EE specification being developed in the open that standardizes the popular Data Repository pattern across a variety of providers. Open Liberty 24.0.0.1-beta includes the Jakarta Data 1.0 Milestone 2 release, which provides API updates to pagination and various improvements to the Javadoc and specification text. The Open Liberty beta includes a test implementation of Jakarta Data that we are using to experiment with proposed specification features so that developers can try out these features and provide feedback to influence the development of the Jakarta Data 1.0 specification beyond Milestone 2. The test implementation currently works with relational databases and operates by redirecting repository operations to the built-in Jakarta Persistence provider.

To use the pagination API to request pages of results in Jakarta Data, you need an Entity and a Repository.

> Start by defining an entity class that corresponds to your data. With relational databases, the entity class corresponds to a database table. The entity properties (public methods and fields of the entity class) generally correspond to the columns of the table. An entity class can be:

Reviewer: ```suggestion
Start by defining an entity class that corresponds to your data. With relational databases, the entity class corresponds to a database table and the entity properties (public methods and fields of the entity class) generally correspond to the columns of the table. An entity class can be:
```

- annotated with `jakarta.persistence.Entity` and related annotations from Jakarta Persistence
- a Java class without entity annotations, in which case the primary key is inferred from an entity property named `id` or ending with `Id` and an entity property named `version` designates an automatically incremented version column.

You define one or more repository interfaces for an entity, annotate those interfaces as `@Repository` and inject them into components via `@Inject`. The Jakarta Data provider supplies the implementation of the repository interface for you.

> Here's a simple entity,

Reviewer: ```suggestion
The following example shows a simple entity:
```

```
@Entity
public class Product {
    @Id
    public long id;

    public String name;

    public float price;

    @Version
    public long version;
}
```

> Here is a repository that defines operations relating to the entity. Your repository interface can inherit from built-in interfaces such as `BasicRepository` and `CrudRepository` to gain a variety of general purpose repository methods for inserting, updating, deleting and querying for entities, but in this case we will define all of the methods ourselves, taking advantage of the life cycle annotations.

Reviewer: ```suggestion
The following example shows a repository that defines operations that relate to the entity. Your repository interface can inherit from built-in interfaces, such as `BasicRepository` and `CrudRepository`, to gain a variety of general purpose repository methods for inserting, updating, deleting, and querying for entities. However, in this case, we will define all of the methods ourselves by using the new lifecycle annotations:
```

```
@Repository(dataStore = "java:app/jdbc/my-example-data")
public interface Products {
    @Insert
    Product add(Product newProduct);

    @Update
    boolean modify(Product product);

    @Delete
    boolean remove(Product product);

    // parameter based query that requires compilation with -parameters to preserve parameter names
    Optional<Product> find(long id);

    // query-by-method name pattern:
    KeysetAwarePage<Product> findByNameIgnoreCaseContains(String searchFor, Pageable pageRequest);

    // query via JPQL:
    @Query("UPDATE Product o SET o.price = o.price - (?2 * o.price) WHERE o.id = ?1")
    boolean discount(long productId, float discountRate);
}
```

> Here is an example of the repository being used,

Reviewer: ```suggestion
The following example shows the repository being used:
```
